### PR TITLE
Move logic to annotate offense message.

### PR DESCRIPTION
**Pull request by melch**
Move logic to annotate offense messages out of the `Cop` class into its own class, `OffenseMessageAnnotater`.

Reasoning for this PR:
* The `Cop` class is close to being too large.
* The moved logic pertains only to annotating the message passed into an offense and not any other part of Cop.
* The description of the cop class is:
> "Cops track offenses and can autocorrect them on the fly."

  Single responsibility principle suggests that cops don't need to also know message annotation logic.

Message annotation was not 100% tested, so attempted to move safely by:
1. Create loose scaffold spec in cop_spec.
2. Move logic into new class.
3. Move existing specs into new class's spec file.
4. Move scaffold specs into new class's spec file.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

**Comment by Drenmi** [comment]
I'm all for this. The new class has a slim public API, and effectively encapsulates a lot of related logic. My only contention would be that Wiktionary considers "annotater" [a typo](https://en.wiktionary.org/wiki/annotater). 😉

**Comment by melch** [comment]
Ha, fair enough 😳 . I'll push a fix to change to "annotator"

File: lib/rubocop/cop/offense_message_annotater.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # A annotater that takes the basic message from a cop
+    # and annotates it for output based on options or settings.
+    class OffenseMessageAnnotater
```
Comment (bbatsov) [inline lib/rubocop/cop/offense_message_annotater.rb]: I'd name this simply `MessageAnnotator`.

File: lib/rubocop/cop/offense_message_annotater.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      attr_reader :options, :config, :cop_config
+
+      def initialize(config, cop_config, options)
+        @config = config
+        @cop_config = cop_config || {}
+        @options = options
+      end
+
+      def annotate_message(message, name)
```
Comment (bbatsov) [inline lib/rubocop/cop/offense_message_annotater.rb]: Probably this can be simply `annotate`.

File: lib/rubocop/cop/offense_message_annotater.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # A annotater that takes the basic message from a cop
+    # and annotates it for output based on options or settings.
```
Comment (bbatsov) [inline lib/rubocop/cop/offense_message_annotater.rb]: Ideally the class documentation should extended a bit and yard doc should be added to the public methods.

**Comment by bbatsov** [comment]
I'm fine with the suggested changed overall, but I've added some remarks that we'll have to be addressed before we can merge this.

**Comment by bbatsov** [comment]
@melch ping :-)

File: lib/rubocop/cop/offense_message_annotater.rb (hunk 3 above)
Comment (melch) [inline lib/rubocop/cop/offense_message_annotater.rb]: I added documentation to public methods and used yard doc syntax. If that is not as much addition to the class documentation, please let me know what other information should be included.

**Comment by melch** [comment]
If these changes look good, please let me know if you are okay merging with squash on your end or if you would like me to squash the commits within the PR.

**Comment by bbatsov** [comment]
Squashed and merged. Thanks!

**Pull request merged**